main.tf

```diff
@@ -147,11 +147,68 @@ resource "metal_device" "esxi_hosts" {
   }
 }
 
+resource "null_resource" "reboot_pre_upgrade" {
+
+  depends_on = [metal_device.esxi_hosts]
+  count      = var.update_esxi ? 1 : 0
+
+  provisioner "local-exec" {
+    command = "sleep 250"
+  }
+}
+
+data "template_file" "upgrade_script" {
+  count    = var.update_esxi ? 1 : 0
+  template = "${file("${path.module}/templates/update_esxi.sh.tpl")}"
+  vars = {
+    esxi_update_filename = "${var.esxi_update_filename}"
+  }
+}
+
+# Run the ESXi update script file in each server.
+# If you make changes to the shell script, you need to update the sed command line number to get rid of te { at the end of the file which gets created by Terraform for some reason.
+resource "null_resource" "upgrade_nodes" {
+
+  depends_on  = [null_resource.reboot_pre_upgrade]
+  count       = var.update_esxi ? length(metal_device.esxi_hosts) : 0
+
+  connection {
+    user        = local.ssh_user
+    private_key = chomp(tls_private_key.ssh_key_pair.private_key_pem)
+    host        = "${element(metal_device.esxi_hosts.*.access_public_ipv4, count.index)}"
+  }
+
+  provisioner "file" {
+    content     = "${data.template_file.upgrade_script.0.rendered}"
+    destination = "/tmp/update_esxi.sh"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "sed -i '27d' /tmp/update_esxi.sh",
+      "echo 'Running update script on remote host.'",
+      "chmod +x /tmp/update_esxi.sh",
+      "/tmp/update_esxi.sh"
+    ]
+  }
+}
+
+resource "null_resource" "reboot_post_upgrade" {
+
+  depends_on = [null_resource.upgrade_nodes]
+  count      = var.update_esxi ? 1 : 0
+
+  provisioner "local-exec" {
+    command = "sleep 250"
```

Comment:
`sleep 250` is going to introduce flakes.

Perhaps, if we can be certain the node has already started the reboot (and SSH is down before this null_resource starts) the following would work:
```tf
resource "null_resource" "reboot_post_upgrade" {
  depends_on = [null_resource.upgrade_nodes]
  count      = var.update_esxi ? length(metal_device.esxi_hosts) : 0
  provisioner "remote-exec" {
    command  = "vmware -vl"
  }
  connection {
    type         = "ssh"
    user         = local.ssh_user
    private_key  = chomp(tls_private_key.ssh_key_pair.private_key_pem)
    timeout      = "250s"
    bastion_host = metal_device.router.access_public_ipv4
    host         = metal_device.esxi_hosts[count.index].access_public_ipv4
    # or we could use the VLAN address and do this after the L2 switch rather than before 
  } 
}
  ```
  
  What are your thoughts here, @empovit ?

Reply:
I'm expecting timeout to act like a retry-timeout here. That may not be the case.

Reply:
@displague you're absolutely right! I didn't really like this part about the [original script](https://github.com/enkelprifti98/packet-esxi-6-7) by Equinix (linked from "Using VMware ESXi on Equinix Metal" which doesn't seem available anymore), but didn't want to introduce too much changes at once - and this procedure looks tried and true. Plus, there are also https://github.com/equinix/terraform-metal-vsphere/commit/25f75e6b10808cb52c0581ce17d6e606e007aff5 and https://github.com/equinix/terraform-metal-vsphere/commit/cd35a4721ed8275807589485aef6924d2e872e4e I'd like to merge at some point.

I've started exploring ways to implement a health check for the two reboots (pre & post upgrade), but haven't come up with usable code yet. IMO ssh-ing to ESXi isn't enough, and the state of internal services must be checked. 

The question is whether you'd like to wait for proper health-checks instead of the sleeps, or is it OK to merge the PR and improve it iteratively, especially that this feature isn't widely used.

Also, if you're rolling out 7.0 support for all machine types (I don't know, badly need it for `g2.large.x86`), then maybe the upgrade feature isn't needed after all. 

WDYT?